File: bindings/csharp/Runtime.cs
using System.Diagnostics;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace MarsRuntime;

public sealed record FeatureSchema(
    [property: JsonPropertyName("n_features")] int? NFeatures,
    [property: JsonPropertyName("feature_names")] string[]? FeatureNames);

public sealed record BasisTerm(
    [property: JsonPropertyName("kind")] string Kind,
    [property: JsonPropertyName("variable_idx")] int? VariableIdx,
    [property: JsonPropertyName("knot_val")] double? KnotVal,
    [property: JsonPropertyName("is_right_hinge")] bool? IsRightHinge,
    [property: JsonPropertyName("category")] JsonElement? Category,
    [property: JsonPropertyName("parent1")] BasisTerm? Parent1,
    [property: JsonPropertyName("parent2")] BasisTerm? Parent2);

public sealed record ModelSpec(
    [property: JsonPropertyName("spec_version")] string SpecVersion,
    [property: JsonPropertyName("params")] JsonElement? Params,
    [property: JsonPropertyName("feature_schema")] FeatureSchema FeatureSchema,
    [property: JsonPropertyName("basis_terms")] BasisTerm[] BasisTerms,
    [property: JsonPropertyName("coefficients")] double[] Coefficients);

public sealed record TrainingParams(
    [property: JsonPropertyName("max_terms")] int MaxTerms,
    [property: JsonPropertyName("max_degree")] int MaxDegree,
    [property: JsonPropertyName("penalty")] double Penalty,
    [property: JsonPropertyName("minspan")] double Minspan,
    [property: JsonPropertyName("endspan")] double Endspan,
    [property: JsonPropertyName("threshold")] double Threshold,
    [property: JsonPropertyName("allow_linear")] bool AllowLinear,
    [property: JsonPropertyName("allow_missing")] bool AllowMissing,
    [property: JsonPropertyName("categorical_features")] int[]? CategoricalFeatures = null,
    [property: JsonPropertyName("feature_names")] string[]? FeatureNames = null);

public sealed record TrainingRequest(
    [property: JsonPropertyName("x")] double[][] X,
    [property: JsonPropertyName("y")] double[] Y,
    [property: JsonPropertyName("sample_weight")] double[]? SampleWeight,
    [property: JsonPropertyName("params")] TrainingParams Params);

public static class Runtime
{
    public static ModelSpec LoadModelSpec(string json)
    {
        var spec = JsonSerializer.Deserialize<ModelSpec>(json)
            ?? throw new InvalidOperationException("malformed artifact: empty model spec");
        Validate(spec);
        return spec;
    }

    public static void Validate(ModelSpec spec)
    {
        if (TryValidateWithRust(spec, out var runtimeError))
        {
            return;
        }

        if (runtimeError is not null)
        {
            throw runtimeError;
        }

        ValidatePure(spec);
    }

    public static double[][] DesignMatrix(ModelSpec spec, double[][] rows)
    {
        if (TryDesignMatrixWithRust(spec, rows, out var matrix, out var runtimeError))
        {
            return matrix;
        }

        if (runtimeError is not null)
        {
            throw runtimeError;
        }

        ValidatePure(spec);
        ValidateRowsPure(spec, rows);
        return rows.Select(row => spec.BasisTerms.Select(term => Evaluate(term, row)).ToArray()).ToArray();
    }

    public static double[] Predict(ModelSpec spec, double[][] rows)
    {
        if (TryPredictWithRust(spec, rows, out var predictions, out var runtimeError))
        {
            return predictions;
        }

        if (runtimeError is not null)
        {
            throw runtimeError;
        }

        return DesignMatrix(spec, rows)
            .Select(row => row.Select((value, idx) => value * spec.Coefficients[idx]).Sum())
            .ToArray();
    }

    public static ModelSpec FitModel(TrainingRequest request)
    {
        if (TryFitWithRust(request, out var spec, out var runtimeError))
        {
            return spec!;
        }

        if (runtimeError is not null)
        {
            throw runtimeError;
        }

        throw new InvalidOperationException("Rust training binary is not available");
    }

    private static bool TryValidateWithRust(ModelSpec spec, out InvalidOperationException? runtimeError)
    {
        runtimeError = null;
        return TryRunRustRuntime("validate", spec, null, out _, out runtimeError);
    }

    private static bool TryDesignMatrixWithRust(
        ModelSpec spec,
        double[][] rows,
        out double[][] matrix,
        out InvalidOperationException? runtimeError)
    {
        matrix = Array.Empty<double[]>();
        runtimeError = null;
        if (!TryRunRustRuntime("design-matrix", spec, rows, out var stdout, out runtimeError))
        {
            return false;
        }

        var payload = JsonSerializer.Deserialize<double?[][]>(stdout)
            ?? throw new InvalidOperationException("Rust runtime returned an empty design matrix");
        matrix = NormalizeMatrix(payload);
        return true;
    }

    private static bool TryPredictWithRust(
        ModelSpec spec,
        double[][] rows,
        out double[] predictions,
        out InvalidOperationException? runtimeError)
    {
        predictions = Array.Empty<double>();
        runtimeError = null;
        if (!TryRunRustRuntime("predict", spec, rows, out var stdout, out runtimeError))
        {
            return false;
        }

        var payload = JsonSerializer.Deserialize<double?[]>(stdout)
            ?? throw new InvalidOperationException("Rust runtime returned empty predictions");
        predictions = NormalizeVector(payload);
        return true;
    }

    private static bool TryFitWithRust(
        TrainingRequest request,
        out ModelSpec? spec,
        out InvalidOperationException? runtimeError)
    {
        spec = null;
        runtimeError = null;
        if (!TryRunRustTraining(request, out var stdout, out runtimeError))
        {
            return false;
        }

        spec = JsonSerializer.Deserialize<ModelSpec>(stdout)
            ?? throw new InvalidOperationException("Rust runtime returned an empty model spec");
        Validate(spec);
        return true;
    }

    private static bool TryRunRustRuntime(
        string command,
        ModelSpec spec,
        double[][]? rows,
        out string stdout,
        out InvalidOperationException? runtimeError)
    {
        stdout = string.Empty;
        runtimeError = null;

        var binary = FindRustRuntimeBinary();
        if (binary is null)
        {
            return false;
        }

        var specPath = WriteTempJson(JsonSerializer.Serialize(spec));
        string? rowsPath = null;
        try
        {
            if (rows is not null)
            {
                rowsPath = WriteTempJson(JsonSerializer.Serialize(ToNullableRows(rows)));
            }

            var psi = new ProcessStartInfo
            {
                FileName = binary,
                RedirectStandardOutput = true,
                RedirectStandardError = true,
                UseShellExecute = false,
                CreateNoWindow = true,
            };
            psi.ArgumentList.Add(command);
            psi.ArgumentList.Add("--spec-file");
            psi.ArgumentList.Add(specPath);
            if (rowsPath is not null)
            {
                psi.ArgumentList.Add("--rows-file");
                psi.ArgumentList.Add(rowsPath);
            }

            using var process = Process.Start(psi);
            if (process is null)
            {
                return false;
            }

            stdout = process.StandardOutput.ReadToEnd();
            var stderr = process.StandardError.ReadToEnd();
            process.WaitForExit();

            if (process.ExitCode != 0)
            {
                runtimeError = new InvalidOperationException(string.IsNullOrWhiteSpace(stderr)
                    ? $"Rust runtime command failed: {command}"
                    : stderr.Trim());
                return false;
            }

            return true;
        }
        catch (Exception ex) when (ex is IOException or UnauthorizedAccessException or System.ComponentModel.Win32Exception)
        {
            return false;
        }
        finally
        {
            TryDelete(specPath);
            if (rowsPath is not null)
            {
                TryDelete(rowsPath);
            }
        }
    }

    private static bool TryRunRustTraining(
        TrainingRequest request,
        out string stdout,
        out InvalidOperationException? runtimeError)
    {
        stdout = string.Empty;
        runtimeError = null;

        var binary = FindRustRuntimeBinary();
        if (binary is null)
        {
            return false;
        }

        var requestPath = WriteTempJson(JsonSerializer.Serialize(request));
        try
        {
            var psi = new ProcessStartInfo
            {
                FileName = binary,
                RedirectStandardOutput = true,
                RedirectStandardError = true,
                UseShellExecute = false,
                CreateNoWindow = true,
            };
            psi.ArgumentList.Add("fit");
            psi.ArgumentList.Add("--request-file");
            psi.ArgumentList.Add(requestPath);

            using var process = Process.Start(psi);
            if (process is null)
            {
                return false;
            }

            stdout = process.StandardOutput.ReadToEnd();
            var stderr = process.StandardError.ReadToEnd();
            process.WaitForExit();

            if (process.ExitCode != 0)
            {
                runtimeError = new InvalidOperationException(string.IsNullOrWhiteSpace(stderr)
                    ? "Rust runtime command failed: fit"
                    : stderr.Trim());
                return false;
            }

            return true;
        }
        catch (Exception ex) when (ex is IOException or UnauthorizedAccessException or System.ComponentModel.Win32Exception)
        {
            return false;
        }
        finally
        {
            TryDelete(requestPath);
        }
    }

    private static string? FindRustRuntimeBinary()
    {
        var envBinary = Environment.GetEnvironmentVariable("MARS_RUNTIME_BIN");
        if (IsExecutableFile(envBinary))
        {
            return envBinary;
        }

        foreach (var start in new[] { Environment.CurrentDirectory, AppContext.BaseDirectory })
        {
            var directory = new DirectoryInfo(Path.GetFullPath(start));
            while (directory is not null)
            {
                foreach (var path in CandidateBinaryPaths(directory.FullName))
                {
                    if (IsExecutableFile(path))
                    {
                        return path;
                    }
                }

                directory = directory.Parent;
            }
        }

        return null;
    }

    private static IEnumerable<string> CandidateBinaryPaths(string root)
    {
        var binaryName = OperatingSystem.IsWindows() ? "mars-runtime-cli.exe" : "mars-runtime-cli";
        yield return Path.Combine(root, "rust-runtime", "target", "debug", binaryName);
        yield return Path.Combine(root, "rust-runtime", "target", "release", binaryName);
    }

    private static bool IsExecutableFile(string? path)
    {
        return !string.IsNullOrWhiteSpace(path) && File.Exists(path);
    }

    private static string WriteTempJson(string content)
    {
        var path = Path.Combine(Path.GetTempPath(), $"{Guid.NewGuid():N}.json");
        File.WriteAllText(path, content);
        return path;
    }

    private static void TryDelete(string path)
    {
        try
        {
            if (File.Exists(path))
            {
                File.Delete(path);
            }
        }
        catch
        {
        }
    }

    private static double?[][] ToNullableRows(double[][] rows)
    {
        return rows.Select(row => row.Select(value => double.IsNaN(value) ? (double?)null : value).ToArray()).ToArray();
    }

    private static double[][] NormalizeMatrix(double?[][] rows)
    {
        return rows.Select(row => NormalizeVector(row)).ToArray();
    }

    private static double[] NormalizeVector(double?[] values)
    {
        return values.Select(value => value ?? double.NaN).ToArray();
    }

    private static void ValidatePure(ModelSpec spec)
    {
        if (string.IsNullOrEmpty(spec.SpecVersion) || spec.SpecVersion.Length < 3 || spec.SpecVersion[1] != '.')
        {
            throw new InvalidOperationException("malformed artifact: spec_version must be '<major>.<minor>'");
        }

        if (!spec.SpecVersion.StartsWith("1.", StringComparison.Ordinal))
        {
            throw new InvalidOperationException($"unsupported artifact version: {spec.SpecVersion}");
        }

        if (spec.BasisTerms.Length != spec.Coefficients.Length)
        {
            throw new InvalidOperationException("malformed artifact: coefficients length must match basis_terms");
        }

        if (spec.FeatureSchema.NFeatures is not null && spec.FeatureSchema.FeatureNames is not null)
        {
            if (spec.FeatureSchema.FeatureNames.Length != spec.FeatureSchema.NFeatures.Value)
            {
                throw new InvalidOperationException("malformed artifact: feature_names length must match n_features");
            }
        }

        for (var index = 0; index < spec.BasisTerms.Length; index++)
        {
            ValidateBasis(spec, spec.BasisTerms[index], index);
        }
    }

    private static void ValidateRowsPure(ModelSpec spec, double[][] rows)
    {
        if (spec.FeatureSchema.NFeatures is null)
        {
            return;
        }

        for (var index = 0; index < rows.Length; index++)
        {
            if (rows[index].Length != spec.FeatureSchema.NFeatures.Value)
            {
                throw new InvalidOperationException(
                    $"feature-count mismatch: row {index} has {rows[index].Length} features, expected {spec.FeatureSchema.NFeatures.Value}");
            }
        }
    }

    private static void ValidateBasis(ModelSpec spec, BasisTerm basis, int index)
    {
        if (string.IsNullOrEmpty(basis.Kind))
        {
            throw new InvalidOperationException($"missing required field: basis term {index} has empty kind");
        }

        switch (basis.Kind)
        {
            case "constant":
                return;
            case "linear":
            case "missingness":
                ValidateVariableIdx(spec, basis.VariableIdx, index);
                return;
            case "hinge":
                ValidateVariableIdx(spec, basis.VariableIdx, index);
                if (basis.KnotVal is null)
                {
                    throw new InvalidOperationException("missing required field: hinge requires knot_val");
                }

                if (basis.IsRightHinge is null)
                {
                    throw new InvalidOperationException("missing required field: hinge requires is_right_hinge");
                }

                return;
            case "categorical":
                ValidateVariableIdx(spec, basis.VariableIdx, index);
                _ = CategoryValue(basis);
                return;
            case "interaction":
                if (basis.Parent1 is null || basis.Parent2 is null)
                {
                    throw new InvalidOperationException("missing required field: interaction requires parent1 and parent2");
                }

                return;
            default:
                throw new InvalidOperationException($"unsupported basis term: {basis.Kind}");
        }
    }

    private static void ValidateVariableIdx(ModelSpec spec, int? variableIdx, int basisIndex)
    {
        if (variableIdx is null)
        {
            throw new InvalidOperationException("missing required field: basis term requires variable_idx");
        }

        if (spec.FeatureSchema.NFeatures is not null && variableIdx.Value >= spec.FeatureSchema.NFeatures.Value)
        {
            throw new InvalidOperationException(
                $"malformed artifact: basis term {basisIndex} references variable outside feature count");
        }
    }

    private static double Evaluate(BasisTerm basis, double[] row)
    {
        return basis.Kind switch
        {
            "constant" => 1.0,
            "linear" => row[basis.VariableIdx!.Value],
            "hinge" => basis.IsRightHinge == true
                ? Math.Max(row[basis.VariableIdx!.Value] - basis.KnotVal!.Value, 0.0)
                : Math.Max(basis.KnotVal!.Value - row[basis.VariableIdx!.Value], 0.0),
            "categorical" => EvaluateCategorical(basis, row),
            "interaction" => EvaluateInteraction(basis, row),
            "missingness" => double.IsNaN(row[basis.VariableIdx!.Value]) ? 1.0 : 0.0,
            _ => throw new InvalidOperationException($"unsupported basis term: {basis.Kind}"),
        };
    }

    private static double EvaluateCategorical(BasisTerm basis, double[] row)
    {
        var category = CategoryValue(basis);
        var value = row[basis.VariableIdx!.Value];
        if (double.IsNaN(value))
        {
            return double.NaN;
        }

        return value == category ? 1.0 : 0.0;
    }

    private static double EvaluateInteraction(BasisTerm basis, double[] row)
    {
        var left = Evaluate(basis.Parent1!, row);
        var right = Evaluate(basis.Parent2!, row);
        return double.IsNaN(left) || double.IsNaN(right) ? double.NaN : left * right;
    }

    private static double CategoryValue(BasisTerm basis)
    {
        if (basis.Category is null || basis.Category.Value.ValueKind is JsonValueKind.Null or JsonValueKind.Undefined)
        {
            throw new InvalidOperationException("missing required field: categorical requires category");
        }

        if (!basis.Category.Value.TryGetDouble(out var value))
        {
            throw new InvalidOperationException("invalid categorical encoding: expected numeric category");
        }

        return value;
    }
}

File: bindings/go/runtime.go
package pymars

import (
	"bytes"
	"encoding/json"
	"errors"
	"fmt"
	"math"
	"os"
	"os/exec"
	"path/filepath"
	"runtime"
	"strings"
)

type FeatureSchema struct {
	NFeatures    *int     `json:"n_features"`
	FeatureNames []string `json:"feature_names"`
}

type BasisTerm struct {
	Kind        string          `json:"kind"`
	VariableIdx *int            `json:"variable_idx"`
	KnotVal     *float64        `json:"knot_val"`
	IsRight     *bool           `json:"is_right_hinge"`
	Category    json.RawMessage `json:"category"`
	Parent1     *BasisTerm      `json:"parent1"`
	Parent2     *BasisTerm      `json:"parent2"`
}

type ModelSpec struct {
	SpecVersion   string        `json:"spec_version"`
	Params        any           `json:"params"`
	FeatureSchema FeatureSchema `json:"feature_schema"`
	BasisTerms    []BasisTerm   `json:"basis_terms"`
	Coefficients  []float64     `json:"coefficients"`
}

type TrainingParams struct {
	MaxTerms            int      `json:"max_terms"`
	MaxDegree           int      `json:"max_degree"`
	Penalty             float64  `json:"penalty"`
	Minspan             float64  `json:"minspan"`
	Endspan             float64  `json:"endspan"`
	Threshold           float64  `json:"threshold"`
	AllowLinear         bool     `json:"allow_linear"`
	AllowMissing        bool     `json:"allow_missing"`
	CategoricalFeatures []int    `json:"categorical_features,omitempty"`
	FeatureNames        []string `json:"feature_names,omitempty"`
}

type TrainingRequest struct {
	X            [][]float64    `json:"x"`
	Y            []float64      `json:"y"`
	SampleWeight []float64      `json:"sample_weight,omitempty"`
	Params       TrainingParams `json:"params"`
}

func LoadModelSpec(raw []byte) (*ModelSpec, error) {
	var spec ModelSpec
	if err := json.Unmarshal(raw, &spec); err != nil {
		return nil, fmt.Errorf("malformed artifact: %w", err)
	}
	if err := Validate(&spec); err != nil {
		return nil, err
	}
	return &spec, nil
}

func Validate(spec *ModelSpec) error {
	if spec == nil {
		return errors.New("malformed artifact: model spec is nil")
	}
	if ok, err := tryValidateWithRust(spec); ok {
		return nil
	} else if err != nil {
		return err
	}
	return validatePure(spec)
}

func DesignMatrix(spec *ModelSpec, rows [][]float64) ([][]float64, error) {
	if ok, matrix, err := tryDesignMatrixWithRust(spec, rows); ok {
		return matrix, nil
	} else if err != nil {
		return nil, err
	}
	return designMatrixPure(spec, rows)
}

func Predict(spec *ModelSpec, rows [][]float64) ([]float64, error) {
	if ok, predictions, err := tryPredictWithRust(spec, rows); ok {
		return predictions, nil
	} else if err != nil {
		return nil, err
	}
	return predictPure(spec, rows)
}

func FitModel(request TrainingRequest) (*ModelSpec, error) {
	payload, available, err := invokeRustTraining(request)
	if !available || err != nil {
		if err != nil {
			return nil, err
		}
		return nil, errors.New("Rust training binary is not available")
	}
	spec, err := LoadModelSpec(payload)
	if err != nil {
		return nil, err
	}
	return spec, nil
}

func tryValidateWithRust(spec *ModelSpec) (bool, error) {
	_, available, err := invokeRustRuntime("validate", spec, nil)
	return available, err
}

func tryDesignMatrixWithRust(spec *ModelSpec, rows [][]float64) (bool, [][]float64, error) {
	payload, available, err := invokeRustRuntime("design-matrix", spec, rows)
	if !available || err != nil {
		return available, nil, err
	}
	var raw [][]*float64
	if err := json.Unmarshal(payload, &raw); err != nil {
		return true, nil, fmt.Errorf("failed to parse Rust runtime output: %w", err)
	}
	return true, runtimeMatrixToFloat(raw), nil
}

func tryPredictWithRust(spec *ModelSpec, rows [][]float64) (bool, []float64, error) {
	payload, available, err := invokeRustRuntime("predict", spec, rows)
	if !available || err != nil {
		return available, nil, err
	}
	var raw []*float64
	if err := json.Unmarshal(payload, &raw); err != nil {
		return true, nil, fmt.Errorf("failed to parse Rust runtime output: %w", err)
	}
	return true, runtimeVectorToFloat(raw), nil
}

func invokeRustRuntime(command string, spec *ModelSpec, rows [][]float64) ([]byte, bool, error) {
	binary := rustRuntimeBinary()
	if binary == "" {
		return nil, false, nil
	}

	specPath, cleanupSpec, ok := writeTempJSON(spec)
	if !ok {
		return nil, false, nil
	}
	defer cleanupSpec()

	var rowsPath string
	var cleanupRows func()
	if rows != nil {
		var okRows bool
		rowsPath, cleanupRows, okRows = writeTempJSON(nullableRows(rows))
		if !okRows {
			return nil, false, nil
		}
		defer cleanupRows()
	}

	args := []string{command, "--spec-file", specPath}
	if rowsPath != "" {
		args = append(args, "--rows-file", rowsPath)
	}

	cmd := exec.Command(binary, args...)
	var stdout bytes.Buffer
	var stderr bytes.Buffer
	cmd.Stdout = &stdout
	cmd.Stderr = &stderr

	if err := cmd.Run(); err != nil {
		if errors.Is(err, exec.ErrNotFound) {
			return nil, false, nil
		}
		var exitErr *exec.ExitError
		if errors.As(err, &exitErr) {
			message := strings.TrimSpace(stderr.String())
			if message == "" {
				message = exitErr.Error()
			}
			return nil, true, errors.New(message)
		}
		return nil, false, nil
	}

	return stdout.Bytes(), true, nil
}

func invokeRustTraining(request TrainingRequest) ([]byte, bool, error) {
	binary := rustRuntimeBinary()
	if binary == "" {
		return nil, false, nil
	}

	requestPath, cleanupRequest, ok := writeTempJSON(request)
	if !ok {
		return nil, false, nil
	}
	defer cleanupRequest()

	args := []string{"fit", "--request-file", requestPath}

	cmd := exec.Command(binary, args...)
	var stdout bytes.Buffer
	var stderr bytes.Buffer
	cmd.Stdout = &stdout
	cmd.Stderr = &stderr

	if err := cmd.Run(); err != nil {
		if errors.Is(err, exec.ErrNotFound) {
			return nil, false, nil
		}
		var exitErr *exec.ExitError
		if errors.As(err, &exitErr) {
			message := strings.TrimSpace(stderr.String())
			if message == "" {
				message = exitErr.Error()
			}
			return nil, true, errors.New(message)
		}
		return nil, false, nil
	}

	return stdout.Bytes(), true, nil
}

func rustRuntimeBinary() string {
	if envBinary := os.Getenv("MARS_RUNTIME_BIN"); envBinary != "" {
		if fileExists(envBinary) {
			return envBinary
		}
	}

	start, err := os.Getwd()
	if err != nil {
		return ""
	}

	for dir := filepath.Clean(start); ; dir = filepath.Dir(dir) {
		for _, candidate := range candidateRuntimeBinaries(dir) {
			if fileExists(candidate) {
				return candidate
			}
		}
		parent := filepath.Dir(dir)
		if parent == dir {
			break
		}
	}

	return ""
}

func candidateRuntimeBinaries(root string) []string {
	name := "mars-runtime-cli"
	if runtime.GOOS == "windows" {
		name += ".exe"
	}
	return []string{
		filepath.Join(root, "rust-runtime", "target", "debug", name),
		filepath.Join(root, "rust-runtime", "target", "release", name),
	}
}

func fileExists(path string) bool {
	info, err := os.Stat(path)
	return err == nil && !info.IsDir()
}

func writeTempJSON(value any) (string, func(), bool) {
	raw, err := json.Marshal(value)
	if err != nil {
		return "", func() {}, false
	}
	file, err := os.CreateTemp("", "mars-runtime-*.json")
	if err != nil {
		return "", func() {}, false
	}
	if _, err := file.Write(raw); err != nil {
		file.Close()
		os.Remove(file.Name())
		return "", func() {}, false
	}
	if err := file.Close(); err != nil {
		os.Remove(file.Name())
		return "", func() {}, false
	}
	return file.Name(), func() { os.Remove(file.Name()) }, true
}

func nullableRows(rows [][]float64) [][]*float64 {
	out := make([][]*float64, len(rows))
	for i, row := range rows {
		out[i] = make([]*float64, len(row))
		for j, value := range row {
			if math.IsNaN(value) {
				continue
			}
			v := value
			out[i][j] = &v
		}
	}
	return out
}

func runtimeMatrixToFloat(rows [][]*float64) [][]float64 {
	out := make([][]float64, len(rows))
	for i, row := range rows {
		out[i] = runtimeVectorToFloat(row)
	}
	return out
}

func runtimeVectorToFloat(values []*float64) []float64 {
	out := make([]float64, len(values))
	for i, value := range values {
		if value == nil {
			out[i] = math.NaN()
		} else {
			out[i] = *value
		}
	}
	return out
}

func validatePure(spec *ModelSpec) error {
	if spec.SpecVersion == "" || len(spec.SpecVersion) < 3 || spec.SpecVersion[1] != '.' {
		return errors.New("malformed artifact: spec_version must be '<major>.<minor>'")
	}
	if !strings.HasPrefix(spec.SpecVersion, "1.") {
		return fmt.Errorf("unsupported artifact version: %s", spec.SpecVersion)
	}
	if len(spec.BasisTerms) != len(spec.Coefficients) {
		return errors.New("malformed artifact: coefficients length must match basis_terms")
	}
	if spec.FeatureSchema.NFeatures != nil && len(spec.FeatureSchema.FeatureNames) > 0 {
		if len(spec.FeatureSchema.FeatureNames) != *spec.FeatureSchema.NFeatures {
			return errors.New("malformed artifact: feature_names length must match n_features")
		}
	}
	for idx := range spec.BasisTerms {
		if err := validateBasis(spec, &spec.BasisTerms[idx], idx); err != nil {
			return err
		}
	}
	return nil
}

func designMatrixPure(spec *ModelSpec, rows [][]float64) ([][]float64, error) {
	if err := validatePure(spec); err != nil {
		return nil, err
	}
	if err := validateRowsPure(spec, rows); err != nil {
		return nil, err
	}
	matrix := make([][]float64, len(rows))
	for i, row := range rows {
		matrix[i] = make([]float64, len(spec.BasisTerms))
		for j := range spec.BasisTerms {
			value, err := evaluateBasis(&spec.BasisTerms[j], row)
			if err != nil {
				return nil, err
			}
			matrix[i][j] = value
		}
	}
	return matrix, nil
}

func predictPure(spec *ModelSpec, rows [][]float64) ([]float64, error) {
	matrix, err := designMatrixPure(spec, rows)
	if err != nil {
		return nil, err
	}
	predictions := make([]float64, len(matrix))
	for i, row := range matrix {
		for j, value := range row {
			predictions[i] += value * spec.Coefficients[j]
		}
	}
	return predictions, nil
}

func validateBasis(spec *ModelSpec, basis *BasisTerm, idx int) error {
	if basis.Kind == "" {
		return fmt.Errorf("missing required field: basis term %d has empty kind", idx)
	}
	switch basis.Kind {
	case "constant":
		return nil
	case "linear", "missingness":
		return validateVariableIdx(spec, basis.VariableIdx, idx)
	case "hinge":
		if err := validateVariableIdx(spec, basis.VariableIdx, idx); err != nil {
			return err
		}
		if basis.KnotVal == nil {
			return errors.New("missing required field: hinge requires knot_val")
		}
		if basis.IsRight == nil {
			return errors.New("missing required field: hinge requires is_right_hinge")
		}
		return nil
	case "categorical":
		if err := validateVariableIdx(spec, basis.VariableIdx, idx); err != nil {
			return err
		}
		_, err := categoryValue(basis)
		return err
	case "interaction":
		if basis.Parent1 == nil || basis.Parent2 == nil {
			return errors.New("missing required field: interaction requires parent1 and parent2")
		}
		return nil
	default:
		return fmt.Errorf("unsupported basis term: %s", basis.Kind)
	}
}

func validateVariableIdx(spec *ModelSpec, variableIdx *int, basisIdx int) error {
	if variableIdx == nil {
		return errors.New("missing required field: basis term requires variable_idx")
	}
	if spec.FeatureSchema.NFeatures != nil && *variableIdx >= *spec.FeatureSchema.NFeatures {
		return fmt.Errorf("malformed artifact: basis term %d references variable outside feature count", basisIdx)
	}
	return nil
}

func validateRowsPure(spec *ModelSpec, rows [][]float64) error {
	if spec.FeatureSchema.NFeatures == nil {
		return nil
	}
	for idx, row := range rows {
		if len(row) != *spec.FeatureSchema.NFeatures {
			return fmt.Errorf("feature-count mismatch: row %d has %d features, expected %d", idx, len(row), *spec.FeatureSchema.NFeatures)
		}
	}
	return nil
}

func evaluateBasis(basis *BasisTerm, row []float64) (float64, error) {
	switch basis.Kind {
	case "constant":
		return 1, nil
	case "linear":
		return row[*basis.VariableIdx], nil
	case "hinge":
		value := row[*basis.VariableIdx]
		if *basis.IsRight {
			return math.Max(value-*basis.KnotVal, 0), nil
		}
		return math.Max(*basis.KnotVal-value, 0), nil
	case "categorical":
		category, err := categoryValue(basis)
		if err != nil {
			return 0, err
		}
		value := row[*basis.VariableIdx]
		if math.IsNaN(value) {
			return math.NaN(), nil
		}
		if value == category {
			return 1, nil
		}
		return 0, nil
	case "interaction":
		left, err := evaluateBasis(basis.Parent1, row)
		if err != nil {
			return 0, err
		}
		right, err := evaluateBasis(basis.Parent2, row)
		if err != nil {
			return 0, err
		}
		if math.IsNaN(left) || math.IsNaN(right) {
			return math.NaN(), nil
		}
		return left * right, nil
	case "missingness":
		if math.IsNaN(row[*basis.VariableIdx]) {
			return 1, nil
		}
		return 0, nil
	default:
		return 0, fmt.Errorf("unsupported basis term: %s", basis.Kind)
	}
}

func categoryValue(basis *BasisTerm) (float64, error) {
	if len(basis.Category) == 0 || string(basis.Category) == "null" {
		return 0, errors.New("missing required field: categorical requires category")
	}
	var value float64
	if err := json.Unmarshal(basis.Category, &value); err != nil {
		return 0, fmt.Errorf("invalid categorical encoding: %w", err)
	}
	return value, nil
}
